File: app/src/main/java/com/elegion/android/ui/base/fragment/BaseFragment.java
package com.elegion.android.ui.base.fragment;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.view.View;

import com.arellomobile.mvp.MvpAppCompatFragment;
import com.elegion.android.ui.base.view.ErrorView;
import com.elegion.android.ui.base.view.LoadingView;
import com.elegion.android.ui.base.view.ToastErrorView;
import com.elegion.android.ui.dialog.LoadingDialog;

import butterknife.ButterKnife;
import butterknife.Unbinder;

/**
 * @author Mike
 */
public class BaseFragment extends MvpAppCompatFragment implements LoadingView, ErrorView {
    protected Unbinder mUnbinder;
    protected boolean mIsAfterOnSavedState;

    private Runnable mRunOnResume;
    protected LoadingView mLoadingView;
    protected ErrorView mErrorView;

    @Override
    public void onViewCreated(final View view, final Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        mIsAfterOnSavedState = false;
        if (mUnbinder == null && view != null) {
            mUnbinder = ButterKnife.bind(this, view);
        }
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        mLoadingView = getLoadingView();
        mErrorView = getErrorView();
    }

    @Override
    public void onResume() {
        super.onResume();
        if (mRunOnResume != null) {
            mRunOnResume.run();
            mRunOnResume = null;
        }
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        mIsAfterOnSavedState = true;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        if (mUnbinder != null) {
            mUnbinder.unbind();
        }
    }

    public void postOnResume(@Nullable Runnable run) {
        if (mIsAfterOnSavedState) {
            mRunOnResume = run;
        } else if (run != null) {
            run.run();
        }
    }

    protected LoadingView getLoadingView() {
        return LoadingDialog.view(getChildFragmentManager());
    }

    protected ErrorView getErrorView() {
        return new ToastErrorView(getActivity());
    }

    @Override
    public void showLoadingIndicator() {
        mLoadingView.showLoadingIndicator();
    }

    @Override
    public void hideLoadingIndicator() {
        mLoadingView.hideLoadingIndicator();
    }

    @Override
    public void showNetworkError() {
        mErrorView.showNetworkError();
    }

    @Override
    public void showUnexpectedError() {
        mErrorView.showUnexpectedError();
    }

    @Override
    public void showErrorMessage(@NonNull String message) {
        mErrorView.showErrorMessage(message);
    }

    @Override
    public void showErrorMessage(@StringRes int message) {
        mErrorView.showErrorMessage(message);
    }

    @Override
    public void hideErrorMessage() {
        mErrorView.hideErrorMessage();
    }
}

File: app/src/main/kotlin/com/elegion/android/bankex/ui/base/fragment/BaseFragment.kt
package com.elegion.android.bankex.ui.base.fragment

import android.os.Bundle
import android.support.annotation.LayoutRes
import android.support.annotation.StringRes
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.arellomobile.mvp.MvpAppCompatFragment
import com.elegion.android.bankex.ui.base.view.ErrorView
import com.elegion.android.bankex.ui.base.view.LoadingView
import com.elegion.android.bankex.ui.base.view.ToastErrorView
import com.elegion.android.bankex.ui.dialog.LoadingDialog

abstract class BaseFragment : MvpAppCompatFragment(), LoadingView, ErrorView {
    protected var isAfterOnSavedState: Boolean = false

    private var runOnResume: Runnable? = null
    protected lateinit var loadingView: LoadingView
    protected lateinit var errorView: ErrorView

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(getLayout(), container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        isAfterOnSavedState = false
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        loadingView = createLoadingView()
        errorView = createErrorView()
    }

    override fun onResume() {
        super.onResume()
        if (runOnResume != null) {
            runOnResume!!.run()
            runOnResume = null
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        isAfterOnSavedState = true
    }

    fun postOnResume(run: Runnable?) {
        if (isAfterOnSavedState) {
            runOnResume = run
        } else {
            run?.run()
        }
    }

    protected fun createLoadingView(): LoadingView = LoadingDialog.view(childFragmentManager)

    protected fun createErrorView(): ErrorView = ToastErrorView(activity!!)

    @LayoutRes
    protected abstract fun getLayout(): Int

    override fun showLoadingIndicator() = loadingView.showLoadingIndicator()

    override fun hideLoadingIndicator() = loadingView.hideLoadingIndicator()

    override fun showNetworkError() = errorView.showNetworkError()

    override fun showUnexpectedError() = errorView.showUnexpectedError()

    override fun showErrorMessage(message: String) = errorView.showErrorMessage(message)

    override fun showErrorMessage(@StringRes message: Int) = errorView.showErrorMessage(message)

    override fun hideErrorMessage() = errorView.hideErrorMessage()
}
